File: app/src/js/presence.js
// La présence Discord — ce que vos amis voient sous votre nom.
//
// Script classique, chargé APRÈS accueil.js et compte.js : il enveloppe
// showPage() et lit profilCourant, qui n'existent qu'après eux.
//
// Deux lignes, pas plus. Discord n'en affiche pas davantage :
//
//     PokéPension
//     Pokédex de Rouge / Bleu        ← ce qu'on fait
//     Tennosei — Aventure 1          ← qui
//     depuis 42 minutes
//
// CE QUI N'Y VA PAS. Une présence se lit par n'importe qui dans la liste
// d'amis, y compris par des gens qu'on ne connaît pas. Le pseudo et le nom
// d'aventure y sont parce qu'ils sont déjà publics dans l'application. Le reste
// n'y est pas : ni l'avancement chiffré — « 3 / 1025 » n'a pas à s'afficher
// devant tout Discord —, ni rien qui vienne de la session.
//
// Côté Rust, presence.rs se tait si Discord n'est pas ouvert. Ici, on ne
// vérifie donc rien : on annonce, et ce qui doit être ignoré l'est.

// Ce qu'on annonce pour chaque écran. Les libellés sont écrits pour être lus
// par quelqu'un d'autre — « Cherche un dresseur » plutôt que « Dresseurs ».
const PRESENCE_ECRANS = {
  home:         'Sur son tableau de bord',
  jeux:         'Choisit un Pokédex',
  dresseurs:    'Cherche un dresseur',
  chasse:       'À la chasse aux chromatiques',
  strategie:    'Prépare une équipe',
  reproduction: 'Consulte la reproduction',
  profil:       'Sur son profil',
};

// Deux mises à jour à quelques millisecondes d'écart ne servent à rien, et
// showPage() est appelée en rafale au démarrage. On laisse retomber.
let presenceMinuteur = null;
const PRESENCE_DELAI = 400;

// ---- Le réglage ------------------------------------------------------------
//
// DANS localStorage, ET NON SUR LE COMPTE. C'est l'inverse du choix fait pour
// « Apparaître dans la liste des dresseurs », et pour une raison précise : cette
// visibilité-là est une donnée de compte, se retirer sur une machine doit valoir
// partout. La présence Discord, elle, dépend de la machine où Discord tourne.
// Quelqu'un peut vouloir l'afficher chez lui et pas au travail, et un réglage
// de compte lui refuserait ce choix.
//
// Même convention que le volume du cri : une clé, une valeur, et le stockage
// refusé n'est jamais fatal.
const PRESENCE_CLE = 'pokearchive-presence-discord';

// Trois états, et le troisième est le plus utile des trois.
//
// « oui »      l'écran, le pseudo et l'aventure — ce qui existait
// « discrete » l'écran seulement
// « non »      rien du tout
//
// POURQUOI UN ENTRE-DEUX. Ce fichier note lui-même qu'une présence se lit par
// n'importe qui dans la liste d'amis, y compris par des gens qu'on ne connaît
// pas. Or les deux lignes ne racontent pas la même chose : « Pokédex de
// Rouge / Bleu » dit ce qu'on FAIT, « Tennosei — Aventure 1 » dit QUI on est.
// Seule la seconde identifie. Couper le tout pour se protéger de la seconde
// revenait à sacrifier la première sans raison.
const PRESENCE_MODES = ['oui', 'discrete', 'non'];

/**
 * Le mode retenu. ÉTEINT tant que personne n'a dit le contraire.
 *
 * LE DÉFAUT A CHANGÉ DE CAMP, et le raisonnement s'est retourné avec lui. Il
 * valait « oui », et le commentaire d'alors disait qu'une valeur abîmée devait
 * retomber sur le défaut plutôt que sur « non », parce que couper la présence
 * de quelqu'un sans qu'il l'ait demandée était le pire des deux maux.
 *
 * C'est l'inverse. Une présence Discord se lit par TOUTE la liste d'amis, y
 * compris par des gens qu'on ne connaît pas : l'annoncer à quelqu'un qui n'a
 * jamais rien demandé est plus grave que de la taire à quelqu'un qui la
 * voulait — le second s'en aperçoit et rallume, le premier ne saura jamais
 * qu'il a été diffusé. Tout ce qui n'est pas un choix explicite vaut donc
 * « non ».
 *
 * Qui a déjà choisi garde son choix : sa valeur est écrite, et elle est lue
 * ici. Seuls ceux qui n'ont jamais ouvert le réglage basculent — et pour
 * eux, précisément, il n'y avait pas de choix à respecter.
 */
function presenceMode(){
  let v = null;
  try{ v = localStorage.getItem(PRESENCE_CLE); }
  catch(e){ /* stockage refusé : on garde le défaut, qui ne montre rien */ }
  return PRESENCE_MODES.indexOf(v) > -1 ? v : 'non';
}

/** Vrai dès qu'on annonce quelque chose, discrètement ou non. */
function presenceActive(){ return presenceMode() !== 'non'; }

/** Ce qu'on fait, en une ligne. */
function presenceQuoi(){
  if(typeof currentPage === 'undefined') return 'Ouvre PokéPension';

  if(currentPage === 'dex' || currentPage === 'home'){
    // Sur un Pokédex de jeu, c'est le jeu qui compte. Sur celui d'ensemble,
    // dire « Pokémon HOME » serait obscur pour qui ne connaît pas.
    const jeu = (typeof gameByKey !== 'undefined') && gameByKey[currentTab];
    if(jeu) return 'Pokédex de ' + (jeu.title || jeu.tab || currentTab);
    if(currentPage === 'home') return PRESENCE_ECRANS.home;
    return 'Sa collection complète';
  }
  return PRESENCE_ECRANS[currentPage] || 'Sur PokéPension';
}

/** Qui, en une ligne. */
function presenceQui(){
  const pseudo = (typeof dresseurCourant !== 'undefined' && dresseurCourant)
    ? dresseurCourant.pseudo : null;
  const aventure = (typeof profilCourant !== 'undefined' && profilCourant)
    ? profilCourant.nom : null;

  if(pseudo && aventure) return pseudo + ' — ' + aventure;
  if(pseudo) return pseudo;
  // Pas encore connecté : on ne laisse pas la ligne vide, Discord afficherait
  // un trou sous le titre.
  return 'Pas encore connecté';
}

/** Annonce l'état courant. Sans effet si Discord n'écoute pas, ou si on l'a coupée. */
function presenceMaj(){
  const invoke = window.__TAURI__ && window.__TAURI__.core && window.__TAURI__.core.invoke;
  if(!invoke) return;                       // banc d'essai, ou page de génération
  if(!presenceActive()) return;
  // Vide en mode discret, et le Rust omet alors la ligne au lieu de l'annoncer
  // creuse — Discord laisserait un trou sous le titre.
  const qui = presenceMode() === 'discrete' ? '' : presenceQui();
  invoke('presence_maj', { etat: { quoi: presenceQuoi(), qui: qui } })
    .catch(function(){ /* une décoration ne remonte pas d'erreur */ });
}

/** Groupe les appels rapprochés : un seul message part. */
function presencePlusTard(){
  if(presenceMinuteur) clearTimeout(presenceMinuteur);
  presenceMinuteur = setTimeout(presenceMaj, PRESENCE_DELAI);
}

/** Efface la présence — à la déconnexion. */
function presenceEffacer(){
  const invoke = window.__TAURI__ && window.__TAURI__.core && window.__TAURI__.core.invoke;
  if(!invoke) return;
  invoke('presence_effacer').catch(function(){});
}

/**
 * Appliquer le choix, tout de suite.
 *
 * COUPER DOIT EFFACER, et non simplement cesser d'annoncer. Sans cet effacement,
 * la dernière présence envoyée resterait affichée sous le nom du joueur pour
 * toute sa liste d'amis, jusqu'à ce qu'il ferme l'application. C'est exactement
 * le défaut qu'avait la déconnexion : le pseudo et l'aventure survivaient au
 * départ, et personne ne s'en apercevait de son côté de l'écran.
 */
function presenceAppliquer(precedent){
  if(!presenceActive()){ presenceEffacer(); return; }

  // EFFACER SEULEMENT SI L'ON VIENT DE CHANGER DE MODE. Discord garde les
  // champs qu'on ne renvoie pas : en passant d'« activée » à « discrète », la
  // seconde ligne — le pseudo et l'aventure — resterait affichée alors qu'on
  // vient justement de demander qu'elle ne le soit plus.
  //
  // Au démarrage, en revanche, il n'y a rien à effacer, et le faire coûtait
  // deux allers-retours de plus à chaque ouverture. Le banc l'a vu : la
  // suppression d'une aventure passait de sept appels à dix.
  if(precedent && precedent !== presenceMode()) presenceEffacer();
  presencePlusTard();
}

// On enveloppe showPage plutôt que d'ajouter un appel à chacune de ses sorties :
// elle en compte trois, et une quatrième ajoutée un jour oublierait la
// présence sans que rien ne le signale. Le procédé est déjà celui de compte.js,
// qui remplace les fonctions parlant au serveur local.
document.addEventListener('DOMContentLoaded', function(){
  if(typeof showPage === 'function'){
    const original = showPage;
    window.showPage = function(){
      const r = original.apply(this, arguments);
      presencePlusTard();
      return r;
    };
  }
  const choix = document.getElementById('presenceMode');
  const etat = document.getElementById('presenceEtat');
  if(choix){
    choix.value = presenceMode();
    const dire = function(){
      if(!etat) return;
      const m = presenceMode();
      etat.textContent =
        m === 'non' ? 'Discord n’affiche plus rien sous ton nom.'
      : m === 'discrete' ? 'Tes amis Discord voient l’écran que tu consultes, '
                         + 'mais ni ton pseudo ni le nom de ton aventure.'
      : 'Tes amis Discord voient l’écran que tu consultes, ton pseudo et ton aventure.';
    };
    dire();
    choix.addEventListener('change', function(){
      const avant = presenceMode();
      const veut = PRESENCE_MODES.indexOf(choix.value) > -1 ? choix.value : 'non';
      try{ localStorage.setItem(PRESENCE_CLE, veut); }
      catch(e){ /* stockage refusé : le choix ne tiendra pas au redémarrage */ }
      dire();
      presenceAppliquer(avant);
    });
    if(typeof syncSelects === 'function') syncSelects();
  }
  presenceAppliquer();
});

File: app/src-tauri/src/presence.rs
//! La présence Discord — ce que vos amis voient sous votre nom.
//!
//! Discord expose un tuyau local : `\\.\pipe\discord-ipc-0` sous Windows, une
//! socket dans /tmp ailleurs. On s'y connecte, on annonce l'identifiant de
//! l'application, et on lui pousse deux lignes de texte. Discord les affiche
//! sous le pseudo, dans la liste des amis.
//!
//! TROIS RÈGLES, et elles tiennent toutes à la même idée : ceci est un
//! ornement, il ne doit jamais gêner.
//!
//! 1. SI DISCORD N'EST PAS LÀ, ON SE TAIT. Pas d'erreur remontée à l'interface,
//!    pas de message. Beaucoup de gens jouent sans Discord ouvert, et
//!    l'application doit leur être parfaitement indifférente.
//!
//! 2. ON NE BLOQUE JAMAIS. Le tuyau peut se fermer quand Discord redémarre. On
//!    tente une reconnexion à la prochaine mise à jour, et on laisse tomber
//!    entre-temps — l'écran n'a pas à attendre après une décoration.
//!
//! 3. RIEN DE PRIVÉ. Le pseudo et le nom d'aventure sont déjà publics dans
//!    l'application ; le reste — le jeton de session, l'identifiant Discord,
//!    l'avancement chiffré — ne sort pas d'ici. Une présence se lit par
//!    n'importe qui dans la liste d'amis.

use discord_rich_presence::{
    activity::{Activity, Assets, Timestamps},
    DiscordIpc, DiscordIpcClient,
};
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

/// L'identifiant de l'application Discord — le même que celui de la connexion.
/// Un identifiant client est public par nature : il voyage dans chaque adresse
/// d'autorisation OAuth. Il se remplace à la compilation, comme l'adresse de
/// l'API, pour qui reprendrait le projet avec sa propre application.
const APPLICATION: &str = match option_env!("POKEPENSION_DISCORD_ID") {
    Some(v) => v,
    None => "1538934470646694030",
};

pub struct Presence {
    client: Mutex<Option<DiscordIpcClient>>,
    /// L'heure d'ouverture, pour que Discord affiche « depuis 42 minutes ».
    /// Fixée une fois : elle compte le temps passé dans l'application, pas
    /// celui passé sur l'écran courant.
    depuis: i64,
}

impl Presence {
    pub fn new() -> Self {
        Self {
            client: Mutex::new(None),
            depuis: SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .map(|d| d.as_secs() as i64)
                .unwrap_or(0),
        }
    }
}

/// Ce que l'interface envoie. Deux lignes, rien de plus — Discord n'en affiche
/// pas davantage, et tout ce qu'on ajouterait serait tronqué sans prévenir.
#[derive(serde::Deserialize)]
pub struct Etat {
    /// Ce qu'on fait : « Pokédex de Rouge / Bleu », « Chasse aux chromatiques ».
    pub quoi: String,
    /// Qui : « Tennosei — Aventure 1 ».
    pub qui: String,
}

/// Pousse l'état, en silence si Discord n'est pas joignable.
///
/// Rend `true` quand l'annonce est passée, sans que l'appelant ait à en faire
/// quoi que ce soit : c'est pour le journal, pas pour l'interface.
#[tauri::command]
pub fn presence_maj(etat: Etat, presence: tauri::State<'_, Presence>) -> bool {
    // Discord tronque en silence au-delà de 128 octets. On coupe nous-mêmes, sur
    // une frontière de caractère : couper un « é » en deux produirait un octet
    // orphelin que Discord refuserait.
    let quoi = tronquer(&etat.quoi, 120);
    let qui = tronquer(&etat.qui, 120);

    let mut garde = match presence.client.lock() {
        Ok(g) => g,
        // Un verrou empoisonné veut dire qu'un fil a paniqué en le tenant. Ce
        // n'est pas une raison pour faire tomber l'application à cause d'une
        // décoration.
        Err(_) => return false,
    };

    if garde.is_none() {
        // `new` ne peut pas échouer dans cette version du crate : il ne fait
        // que retenir l'identifiant. C'est `connect` qui va chercher le tuyau,
        // et c'est lui qui échoue quand Discord n'est pas ouvert.
        let mut c = DiscordIpcClient::new(APPLICATION);
        if c.connect().is_err() {
            return false; // Discord fermé : c'est le cas ordinaire, pas une erreur.
        }
        *garde = Some(c);
    }

    let client = match garde.as_mut() {
        Some(c) => c,
        None => return false,
    };

    // La seconde ligne est FACULTATIVE, et c'est ce qui rend la presence
    // discrete possible. `qui` porte le pseudo et le nom d'aventure — la seule
    // partie qui identifie. En mode discret, l'interface l'envoie vide, et une
    // chaine vide posee en `state` laisserait un trou sous le titre : c'est
    // justement ce que presence.js evite en ecrivant « Pas encore connecte »
    // plutot que rien. On omet donc la ligne au lieu de l'annoncer creuse.
    let mut activite = Activity::new()
        .details(&quoi)
        .assets(
            Assets::new()
                // La clé de l'image se déclare dans le portail Discord, onglet
                // Rich Presence > Art Assets. Absente, Discord n'affiche
                // simplement pas d'image — rien ne casse.
                .large_image("pokepension")
                .large_text("PokéPension"),
        )
        .timestamps(Timestamps::new().start(presence.depuis));

    if !qui.is_empty() {
        activite = activite.state(&qui);
    }

    if client.set_activity(activite).is_err() {
        // Le tuyau s'est fermé — Discord a redémarré, ou s'est fermé pendant
        // qu'on écrivait. On oublie le client : la prochaine mise à jour
        // rouvrira. Retenter ici ferait attendre l'interface pour rien.
        *garde = None;
        return false;
    }
    true
}

/// Efface la présence. Appelée quand on se déconnecte : laisser « Tennosei —
/// Aventure 1 » affiché après une déconnexion serait au mieux étrange.
#[tauri::command]
pub fn presence_effacer(presence: tauri::State<'_, Presence>) {
    if let Ok(mut garde) = presence.client.lock() {
        if let Some(c) = garde.as_mut() {
            let _ = c.clear_activity();
        }
    }
}

fn tronquer(s: &str, max: usize) -> String {
    if s.len() <= max {
        return s.to_string();
    }
    let mut fin = max;
    while fin > 0 && !s.is_char_boundary(fin) {
        fin -= 1;
    }
    s[..fin].trim_end().to_string()
}
